tensorflow/python/kernel_tests/conv_ops_3d_test.py

```diff
@@ -45,22 +47,34 @@ def GetTestConfigs():
 
 class Conv3DTest(test.TestCase):
 
+  def _DtypesToTest(self, use_gpu):
+    if use_gpu:
+      if not test_util.CudaSupportsHalfMatMulAndConv():
+        return [dtypes.float32]
+      else:
+        # It is important that float32 comes before float16 here,
+        # as we will be using its gradients as reference for fp16 gradients.
+        return [dtypes.float32, dtypes.float16]
+    else:
+      return [dtypes.float64, dtypes.float32, dtypes.float16]
+
   def _SetupValuesForDevice(self, tensor_in_sizes, filter_in_sizes, stride,
-                            padding, data_format, use_gpu):
+                            padding, data_format, dtype, use_gpu):
     total_size_1 = 1
     total_size_2 = 1
     for s in tensor_in_sizes:
       total_size_1 *= s
     for s in filter_in_sizes:
       total_size_2 *= s
 
-    # Initializes the input tensor with array containing incrementing
-    # numbers from 1.
-    x1 = [f * 1.0 for f in range(1, total_size_1 + 1)]
-    x2 = [f * 1.0 for f in range(1, total_size_2 + 1)]
+    # Initializes the input tensor with array containing numbers from 0 to 1.
+    # We keep the input tensor values fairly small to avoid overflowing a float16 
+    # tensor during the conv3d and to keep absolute errors small
```

Comment:
Why does it matter that the absolute error is small? If the absolute error were larger, you can adjust what you pass to atol in self.assertAllClose.

Also, I just realized that you didn't modify what you pass to atol in self.assertAllClose. Can you make it a bit smaller to take into account the expected output values are smaller?

Reply:
>Why does it matter that the absolute error is small?

Because if I understand correctly we compare results with reference using absolute difference. And absolute error depend on current significant and exponent. Large numbers (huge exponent) and small amount of bit for significant (fp16) => big absolute errors (I seen difference >10 with huge numbers in some fp16 tests).

I can make 'tolerances' smaller but not sure which values are appropriate. It's hard to tell because convs are not 'pure convs' but some tricky methods from computation mathematics. Error depends on particular implementation of convolution and if we want to be really sure it's better to find out what are relative/absolute errors for particular method and particular type. But I think that from practical point of view it's better to take 'small enough' values and continue to work on other problems.

Reply:
Sorry for late response. I see your point about making 'tolerances' smaller, so you can leave atol as is.

I think you should remove the "and to keep absolute errors small" comment. You're right that having a large exponent with fp16 means large absolute errors, but I don't see that as a problem. We could have just raised 'atol' in that case. Differences greater than 10 seem reasonable given how imprecise fp16 is. However, currently, some expected_outputs are above the maximum fp16 value of 65504, which would cause them to overflow. This seems like the only reason to reduce the numbers in the input tensor IMO.